Complete blog example

keshon opened the pull request:
I made blog example complete. Few comments about changes:
- Added missing routes.
- Rewritten templates (old are gone).
- Small changes in sql scheme.
- Different visual style (using Marx classless framework with some custom touches) .

Few things to consider:
- I have almost zero experience in Nim.
- I'm not entirely familiar with code guidelines (and I use 4 spaces instead of 2).

xflywind commented:
Thanks for your effores.
Please use `nimpretty --indent:2` to change four space to two space.
```
C:\WINDOWS\system32>nimpretty --help
nimpretty - Nim Pretty Printer Version 0.2
  (c) 2017 Andreas Rumpf
Usage:
  nimpretty [options] file.nim
Options:
  --out:file            set the output file (default: overwrite the input file)
  --indent:N[=0]        set the number of spaces that is used for indentation
                        --indent:0 means autodetection (default behaviour)
```

xflywind commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+*.exe
```
Please add `.gitignore` to .ignore files.

keshon replied:
I added gitignore to itself but I don't know how to remove it from my github repo.
I used nimpretty on all my files.

xflywind replied:
Ok. I will delete them.

xflywind commented:
Thanks! The code looks good to me.

xflywind commented:
Maybe you can explain how to run the projects in readme.(first initialzie the database or so) `not an error` is confusing for beginner. :)

xflywind commented:
![image](https://user-images.githubusercontent.com/43030857/93019853-3a842280-f60c-11ea-8143-457d5cac9d86.png)

I ran into some errors. Do you have some ideas?

xflywind commented:
I have tried it. The project looks really nice. I will look into some codes now.

keshon commented:
Please could you try to run the example in Incognito/private mode to be sure old cookies don't conflict with the new one?С уважением, Иннокентий. ---- On Вс, 13 сен 2020 16:59:01 +0300  notifications@github.com  wrote ----

I ran into some errors. Do you have some ideas?

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

keshon commented:
I can extend readme file to give better understanding. Give me few days.The db init is automatic: if db file doesn't exist it will be created automatically on first request. I will reflect it in readme.С уважением, Иннокентий. ---- On Вс, 13 сен 2020 16:54:34 +0300  notifications@github.com  wrote ----
Maybe you can explain how to run the projects in readme.(first initialzie the database or so) not an error is confusing for beginner. :)

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

keshon commented:
I will take a look at it as well. I remember this error but don't remember how I managed to fix it. Just got away from PC already.С уважением, Иннокентий. ---- On Вс, 13 сен 2020 17:06:37 +0300  notifications@github.com  wrote ----
I have tried it. The project looks really nice. I will look into some codes now.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

xflywind commented:
All right. I think I have solved that problem.

keshon commented:
> All right. I think I have solved that problem.

If I'm not mistaken this error might occur because you've probably used old blog.db database file. But since you've resolved it already it doesn't matter :)

xflywind commented on examples/blog/initdb.nim:
```diff
@@ -3,14 +3,14 @@ import db_sqlite, os, strutils, logging
+      db = open(consts.dbPath, "", "", "")
+      schema = readFile(schemaPath)
+    for line in schema.split(";"):
+      if line == "\c\n":
```
I think this line should be changed to this, or it cause errors when first initial database. You can delete db file to reproduce it.
```nim
if line == "\c\n" or line == "\n":
   continue
```

xflywind replied:
d:\QQPCMgr\Desktop\tp\examples\blog>nim c -r app.nim
Hint: used config file 'C:\Users\blue\.choosenim\toolchains\nim-1.2.6\config\nim.cfg' [Conf]
Hint: used config file 'd:\QQPCMgr\Desktop\tp\config.nims' [Conf]
Hint: used config file 'd:\QQPCMgr\Desktop\tp\examples\blog\config.nims' [Conf]
Hint: 15808 LOC; 0.040 sec; 17.852MiB peakmem; Debug build; proj: d:\QQPCMgr\Desktop\tp\examples\blog\app.nim; out: d:\QQPCMgr\Desktop\tp\examples\blog\app.exe [SuccessX]
Hint: d:\QQPCMgr\Desktop\tp\examples\blog\app.exe  [Exec]
d:\QQPCMgr\Desktop\tp\examples\blog\app.nim(8) app
d:\QQPCMgr\Desktop\tp\examples\blog\initdb.nim(16) initDb
C:\Users\blue\.choosenim\toolchains\nim-1.2.6\lib\impure\db_sqlite.nim(207) exec
C:\Users\blue\.choosenim\toolchains\nim-1.2.6\lib\impure\db_sqlite.nim(142) dbError
Error: unhandled exception: not an error [DbError]
Error: execution of an external program failed: 'd:\QQPCMgr\Desktop\tp\examples\blog\app.exe '

keshon replied:
I added your suggestion but still wasn't able to reproduce strange behavior you have. Did you test on 1.2.6 nim?
I also fixed DROP statement in schema file (i made a typo).

keshon commented:
Ok I will take a look tomorrow.С уважением, Иннокентий. ---- On Вс, 13 сен 2020 17:51:41 +0300  notifications@github.com  wrote ----
@xflywind commented on this pull request.



In examples/blog/initdb.nim:
> -  let
-    db = open(consts.dbPath, "", "", "")
-    schema = readFile(schemaPath)
-  for line in schema.split(";"):
-    if line == "\c\n":
-      continue
-    db.exec(sql(line.strip))
-  db.close()
-  logging.info("Initialized the database.")
+proc initDb*() =
+  if not fileExists(consts.dbPath):
+    let
+      db = open(consts.dbPath, "", "", "")
+      schema = readFile(schemaPath)
+    for line in schema.split(";"):
+      if line == "\c\n":

d:\QQPCMgr\Desktop\tp\examples\blog>nim c -r app.nim
Hint: used config file 'C:\Users\blue.choosenim\toolchains\nim-1.2.6\config\nim.cfg' [Conf]
Hint: used config file 'd:\QQPCMgr\Desktop\tp\config.nims' [Conf]
Hint: used config file 'd:\QQPCMgr\Desktop\tp\examples\blog\config.nims' [Conf]
Hint: 15808 LOC; 0.040 sec; 17.852MiB peakmem; Debug build; proj: d:\QQPCMgr\Desktop\tp\examples\blog\app.nim; out: d:\QQPCMgr\Desktop\tp\examples\blog\app.exe [SuccessX]
Hint: d:\QQPCMgr\Desktop\tp\examples\blog\app.exe  [Exec]
d:\QQPCMgr\Desktop\tp\examples\blog\app.nim(8) app
d:\QQPCMgr\Desktop\tp\examples\blog\initdb.nim(16) initDb
C:\Users\blue.choosenim\toolchains\nim-1.2.6\lib\impure\db_sqlite.nim(207) exec
C:\Users\blue.choosenim\toolchains\nim-1.2.6\lib\impure\db_sqlite.nim(142) dbError
Error: unhandled exception: not an error [DbError]
Error: execution of an external program failed: 'd:\QQPCMgr\Desktop\tp\examples\blog\app.exe '

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

xflywind commented on examples/blog/views.nim:
```diff
@@ -54,22 +63,83 @@ proc register*(ctx: Context) {.async.} =
+  let db = open(consts.dbPath, "", "", "")
+  var posts: seq[seq[string]] = @[]
+  for x in db.fastRows(sql"SELECT * FROM posts"):
+    echo x
```
comment echo or use logging

xflywind commented:
I have reviewed all the codes. They seem to work fine. I think it needs a simple description or a sceenshoot in  readme before merging. Then you could post the readme in [reddit](https://www.reddit.com/r/nim) or not. You can also add your name in readme such as (Designed by @keshon). Thanks!

keshon commented:
> I have reviewed all the codes. They seem to work fine. I think it needs a simple description or a sceenshoot in readme before merging. Then you could post the readme in [reddit](https://www.reddit.com/r/nim) or not. You can also add your name in readme such as (Designed by @keshon). Thanks!

I updated Readme.md take a look at it please. English is not my native to be honest so you are free to edit the file.
I did make a simple screenshot but dont know how to attach it to Readme.md properly (I mean I don't know where do I store the image). Attaching the screenshot here.
![preview](https://user-images.githubusercontent.com/13551109/93059278-29252f80-f679-11ea-8dc6-eb5cbeeeae90.jpg)

As for reddit - I'm kinda speechless guy when it comes to audience, will think about it.

xflywind commented:
Amazing work! Thanks!
